### PR TITLE
Minimum risk trainer (with optional DA)

### PR DESCRIPTION
**Pull request by StoyanVenDimitrov**

### CODE REVIEW

File: neuralmonkey/learning_utils.py [hunk 1]
```diff
@@ -192,8 +192,8 @@ def training_loop(tf_manager: TensorFlowManager,
                         train=True)
                     last_log_time = time.process_time()
                 else:
-                    tf_manager.execute(batch_dataset, [trainer],
-                                       train=True, summaries=False)
+                    log(tf_manager.execute(batch_dataset, [trainer],
```
Comment (jlibovicky) [inline neuralmonkey/learning_utils.py]: Isn't it just a forgotten debug print?

**Comment by StoyanVenDimitrov** [comment]
Hi Jindřich,

I tested it on the data and with the same pretrained model the bandit trainer was tested. But it was slow, it took 80 hours for 15,7k train training examples on a CPU cluster with small batches of 3 and really small sample size of 10 (compared to 80 and 100 respectively by Shen et al.). Still, it improved continuously on the validation set, reaching 30.50 BLEU after a couple of validations, compared to 30.03 of the MLE pretrained model. The bandit learning couldn't make improvements on the same task. So it seems to work, but without a GPU (cluster) it´s very slow. 
About the tests, excuse me for throwing it like this, I will make the changes you suggest asap. I also didn't write any unit tests, because I don't know how they should look like. Do you have any examples?

Best,
Stoyan

**Comment by jindrahelcl** [comment]
Hi Stoyan, we don't have many unit tests implemented so don't bother with them unless you really want to :-) Example tests are in `neuralmonkey/tests/` (not to be confused with `tests/`).

**Comment by varisd** [comment]
Toto muzem zavrit, az na ten annealing je to vsechno vyresene v #720 (a ten annealing neni IMHO nic prevratneho)

**Comment by jindrahelcl** [comment]
A nemohli bychom to odtud vykuchat?

**Comment by varisd** [comment]
Mohli. Jde spis o to, ze nemam zdroj, podle ktereho bych mohl overit, ze je ta metoda spravne implementovana.

**Comment by jindrahelcl** [comment]
So let's open up a feature issue with the annealing, link this pull request
in it and close this.

Dne čt 12. 7. 2018 20:54 uživatel Dušan Variš <notifications@github.com>
napsal:
